app/code/core/Mage/Catalog/Model/Category.php

```diff
@@ -918,11 +918,17 @@ public function getAvailableSortByOptions() {
      * @return string
      */
     public function getDefaultSortBy() {
-        if (!$sortBy = $this->getData('default_sort_by')) {
+
+        $sortBy = $this->getData('default_sort_by');
+        $available = $this->getAvailableSortByOptions();
+
+        //when not set or not available use default from system config
+        if (!$sortBy || !isset($available[$sortBy])) {
             $sortBy = Mage::getSingleton('catalog/config')
                 ->getProductListDefaultSortBy($this->getStoreId());
         }
-        $available = $this->getAvailableSortByOptions();
+
+        //even the sort from system config not available, use the first of available
         if (!isset($available[$sortBy])) {
```

Comment:
Doesn't this `if` block basically clobber the effects of the first `if` block when `!isset($available[$sortBy])` is true?

Reply:
no, the current logic is:

1. try sort from category object; cames from: `$this->getData('default_sort_by')`
2. sort from category object is empty or not available, try use sort from system config; cames from: `Mage::getSingleton('catalog/config')->getProductListDefaultSortBy($this->getStoreId());`
3. sort from system config is also not available, use the first available; cames from `$sortBy[0]`


In the original logic was the sort from system config used only when category sort was empty.

Reply:
The current logic can be simplified like so:

```php
if ($a || $b) {
   $c = 1;
}
if ($b) {
  $c = 2;
}
```

This is equivalent to:

```php
if ($a) {
  $c = 1;
}
if ($b) {
  $c = 2;
}
```

Reply:
I didn't catch your hint. Show me example on actually the method without changing the logic:

```php
    public function getDefaultSortBy() {

        $sortBy = $this->getData('default_sort_by');
        $available = $this->getAvailableSortByOptions();

        //when not set or not available use default from system config
        if (!$sortBy || !isset($available[$sortBy])) {
            $sortBy = Mage::getSingleton('catalog/config')
                ->getProductListDefaultSortBy($this->getStoreId());
        }

        //even the sort from system config not available, use the first of available
        if (!isset($available[$sortBy])) {
            $sortBy = array_keys($available);
            $sortBy = $sortBy[0];
        }

        return $sortBy;
    }
```

Reply:
My bad, I misunderstood the code.. So hard to read with comments inline.. :)

I think all I'd change is to add `! isset($available[$sortBy])` to the last `if` in case the system config value is not an available option.

```php
    public function getDefaultSortBy() {
        $sortBy = $this->getData('default_sort_by');
        $available = $this->getAvailableSortByOptions();
        if ( ! $sortBy || ! isset($available[$sortBy])) {
            $sortBy = Mage::getSingleton('catalog/config')
                ->getProductListDefaultSortBy($this->getStoreId());
        }
        if ( ! $sortBy || ! isset($available[$sortBy])) {
            $sortBy = array_shift($available);
        }
        return $sortBy;
    }
```

Reply:
fixed now with your hint